doc/pages/faq.asciidoc

```diff
@@ -101,97 +206,104 @@ However, you can easily declare key mappings in your configuration file
+cases. You might consider writing a custom `$HOME/.magic` file if needed.
 
-== Can I disable syntax highlighting completely ?
+=== Can I disable syntax highlighting completely?
 
 Similarly to the indentation hooks, the name format followed by the
 highlighting hooks is `<lang>-highlight`. You can thus disable syntax
 highlighting using the following command: `set global disabled_hooks
 '.+-highlight'`
 
-== Why does a dot `.` in a regex select newline characters ?
+== The editing language
+
+=== Why aren't there other scopes similar to `%sh{}` e.g. python?
+
+Supporting custom scopes would add hard dependencies to the project, which
+is too much of a drawback when balanced against the low cost of using
+an interpreter in a regular shell scope (e.g. `%sh{ python -c "..." }`).
+The shell scope allows users to spawn any interpreter they want, for a minimal
+cost in terms of performance, it is therefore the reason why it's the only
+one available by default.
+
+=== What shell is used to expand `%sh{}` scopes?
+
+The server expands shell scopes using the `sh` binary, located in one of
+the directories containing all the POSIX standard utilities. This list of
+directories is stored in a system configuration variable, and queried by
+Kakoune at startup.
+
+In most distributions, `/bin/sh` will end up being used.
+
+=== Why does a dot `.` in a regex select newline characters?
 
 Data in buffers is a stream of characters, and newlines do not receive special
 treatment compared to other characters, with regards to regex matching. In
 order to select data in a line without any trailing newline characters, one could
 use the `[^\n]+` pattern, which is arguably a good compromise when
 balanced against the ability to select data over several lines.
 
-== Can I split the window to display different buffers in them ?
+You can instruct the regex engine to stop matching newline characters with
+`.` by disabling the appropriate flag (`(?S)`).
 
-As a fairly compliant follower of the UNIX philosophy, Kakoune does not
-try to implement features that are best handled by separate, dedicated
-tools. Window splitting in terminals is a prime example of that
-concept, where the editor provides commands to interact with several
-terminal multiplexers (e.g. `tmux`), as opposed to emulating their
-functionalities.
-
-In order to open buffers in the same window simultaneously using `tmux`
-(or one of the supported multiplexers), run Kakoune in a `tmux` session,
-and simply use the `:new` command to spawn new clients as you would
-have otherwise in an X11 environment.
-
-== Why does `a` extend the current selection, but `i` leaves it untouched ?
+=== Why does `a` extend the current selection, but `i` leaves it untouched?
 
 Selections are ranges of characters whose delimiters are an "anchor" and
-a "cursor", and inserting characters is always done before the cursor in
+a "cursor", and inserting characters is always done before the anchor in
```

Comment:
I believe this should remain "cursor"

Reply:
I changed that because when a user has a left-to-right selection and wonders where new text is going inserted if they hit <kbd>i</kbd>, the text will be inserted before the anchor - even if technically, <kdb>i</kbd> swaps cursor/anchor around at the moment you hit it.

The problem with this sentence is that it assumes the selections are left-to-right oriented. If the selection is right-to-left oriented, <kbd>i</kbd> won't move the cursor.

I think we should rewrite this bit entirely to take into account both the above issues. I don't know how, it challenges my skills at English.

Reply:
If I understand correctly, the real answer to this FAQ is something like:

- A selection is a range of characters from the "cursor" (the end you move around directly, sometimes called the head) to the "anchor" (the end that follows the previous position of the cursor, sometimes called the tail)
- Kakoune has (somewhat arbitrarily) decided that inserted text always appears to the left of the cursor
- When you enter insert mode, Kakoune ensures that each selection's cursor is at the left-hand end of the selection — because inserted text appears to the left of the cursor, it appears to the left of the selection, and is therefore not part of the selection
- When you enter append mode, Kakoune ensures that each selection's cursor is at the right-hand end of the selection, and grows the selection by one character — because inserted text appears to the left of the cursor, it appears just inside the selection and is therefore part of the selection

I don't know if there's a specific reason why inserted text always appears to the left of the cursor.

Reply:
@Screwtapello I am not sure what other desirable alternative we would have:

* We could insert *after* the cursor in append mode then move the cursor by the inserted amount, thats quite equivalent to what we do now except that the cursor is always onto the last inserted character and it is a bit more stateful (insert mode behaves differently depending how we entered it).

* We could insert in front of the selection/past the end of the selection and grow it, effectively using a hidden cursor concept that gets decoupled from the selection cursor.

Not growing the selection on append would be very strange, and that would mean the inserted text progressively goes away from the selected text, requiring an additional visual cue on where the text gets inserted, this is not an issue on insert because the selection naturally moves due to the inserted text.

Did you have something specific in mind ?

Reply:
Sorry, I didn't mean to imply that there was some obvious better alternative that you'd missed. Kakoune's selections are minimum-length-1 for reasons I'm confident I can explain, which means that a newly-inserted character must go to the left or the right — but I wasn't confident I could explain why either choice was better than the other.

Now that I've had time to think about it a little more, I think the *real* source of Kakoune's left-bias is the fact that it's designed for Latin script (like terminals and the QWERTY keyboard layout), which is written left-to-right. So, when you insert text, the left-end of the selection moves right to make room for it. When you append text, the right-end of the selection moves right to make room for it. Kakoune's behaviour is mechanically consistent, it's just not visually consistent because of the way the Latin script works. So, here's another attempt at answering this FAQ:

> A selection is a range of characters from the "cursor" (the end you move around directly, sometimes called the head) to the "anchor" (the end that follows the previous position of the cursor, sometimes called the tail). When you type a letter in insert mode, Kakoune will always place it under the cursor (shifting any existing text to the right to make room), and then move the cursor to the right ready for the next letter.
>
> When you enter insert mode with `i`, Kakoune will ensure that each selection's cursor is at the left-hand end before you start typing, so that the text you type appears before the originally selected text. As you type, the cursor (and any following text) move right to make room, leaving the new text outside the selection.
>
> When you enter insert mode with `a`, Kakoune will ensure each selection's cursor is at the right-hand end before you start typing, so that the text you type appears after the originally selected text. As you type, the cursor (and any following text) move right to make room, leaving the new text inside the selection.
>
> Although it's visually inconsistent for appended text to wind up inside the selection and inserted text to be outside it, it's mechanically consistent.

Reply:
No need to be sorry, its actually good to challenge the design and make sure we have good reasons for making things work the way they do.

Regarding the description proposal, in my mind a selection goes from the anchor to the cursor rather than the other way around because the anchor is what might remain fixed (as its name implies) and the cursor is the one moving.

Reply:
So, what do we do with the hunk?

Reply:
Can we restore cursor for now until we find a better wording, "inserting characters is always done before the cursor in insert mode" might be confusing, but is not wrong, if Kakoune is in insert mode, regardless of how it got there, characters will be inserted before the cursor.